### PR TITLE
wip: Introduce signing of OCI artifacts and refactoring

### LINKED ISSUE #835: Introduce signing of OCI artifacts
Part 1 of #722.

### PR DESCRIPTION
**Pull request by maxgio92**
This PR aims to introduce initial support for signing of OCI artifacts leveraged by the [Falco Open Infra](https://github.com/falcosecurity/test-infra).

In order to introduce this feature it also includes refactoring and cleanup work.

Fixes #835

### CODE REVIEW

File: images/build-libs/Makefile [hunk 1]
```diff
@@ -30,3 +34,14 @@ push-image: create-builder
 
 local-registry: create-builder
 	docker buildx build --push --platform $(ARCHS) -t "localhost:5000/$(IMG_NAME)" .
+
+sign-image: cosign
+	$(cosign) sign --key $(cosign_key) $(IMAGE)
+
+cosign: go
```
Comment (developer-guy) [inline images/build-libs/Makefile:41]: maybe we should replace this target name with "install-cosign," wdyt?

File: images/docker-dind/Dockerfile [hunk 2]
```diff
@@ -41,3 +46,10 @@ RUN tmpdir=`mktemp -d` && curl "https://awscli.amazonaws.com/awscli-exe-linux-x8
 	unzip $tmpdir/awscliv2.zip -d $tmpdir && \
 	$tmpdir/aws/install -i /usr/local/aws-cli -b /usr/local/bin && \
 	rm -r $tmpdir
+
+RUN apt-get update && TMPDIR=$(mktemp -d) \
```
Comment (developer-guy) [inline images/docker-dind/Dockerfile:50]: I can recommend you a better way of installing cosign into an image (optional):

> https://github.com/sigstore/cosign#dockerfile

Comment (maxgio92) [reply images/docker-dind/Dockerfile:50]: Good idea multi-stage here! thanks @developer-guy

Comment (maxgio92) [reply images/docker-dind/Dockerfile]: Applied, thanks!

File: images/build-libs/Makefile (hunk 1 above)
Comment (maxgio92) [reply images/build-libs/Makefile:41]: I think that as Cosign is a tool, is a tool target to be satisfied as a dependency, but it's only about names :)

**Comment by maxgio92** [comment]
This PR misses the integration with external vault.

**Comment by developer-guy** [comment]
We'll take a look at this today w/@dentrax to check whether something exists that we can help.

**Comment by maxgio92** [comment]
Please take a look at the motivations [here](https://github.com/falcosecurity/test-infra/issues/722#issuecomment-1236713048) @developer-. If you have feedbacks they're super welcome :)

File: images/build-libs/Makefile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+# To avoid using 'docker manifest' (that is experimental), 
+# allow buildx to manage push to registry for us.
+# buildx manages build and push in one step
+build-push: push-image
+
+create-builder:
+	# Workaround for buildx issue, see https://github.com/docker/buildx/issues/495#issuecomment-761562905
+	docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+	docker buildx create --name libsbuilder --driver docker-container --use
```
Comment (developer-guy) [inline images/build-libs/Makefile:26]: I think you could use "--bootstrap" flag in here, no need an extra command for this.

File: images/build-plugins/Makefile [hunk 4]
```diff
@@ -1,4 +1,8 @@
-SHELL := /bin/bash
+SHELL := /usr/bin/env bash
+
+cosign := $(shell command -v cosign 2>/dev/null)
+cosign_version ?= 1.9.0
```
Comment (developer-guy) [inline images/build-plugins/Makefile:4]: you might use 1.11.1, which is the latest at the time of writing this.

File: images/common.Makefile [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+bins := docker aws cosign go
+
+# OCI build
+BUILD_CONTEXT ?= .
+DOCKERFILE ?= Dockerfile
+CACHE_FLAG ?= --no-cache
+
+# OCI artifact signing
+cosign_version ?= 1.9.0
```
Comment (developer-guy) [inline images/common.Makefile:11]: 1.11.1

File: images/docker-dind/Dockerfile [hunk 6]
```diff
@@ -1,5 +1,12 @@
+ARG COSIGN_VERSION=v1.9.0
+
+FROM gcr.io/projectsigstore/cosign:${COSIGN_VERSION} as cosign-bin
```
Comment (developer-guy) [inline images/docker-dind/Dockerfile:3]: nice move !

File: images/docker-dind/Dockerfile [hunk 7]
```diff
@@ -41,3 +48,7 @@ RUN tmpdir=`mktemp -d` && curl "https://awscli.amazonaws.com/awscli-exe-linux-x8
 	unzip $tmpdir/awscliv2.zip -d $tmpdir && \
 	$tmpdir/aws/install -i /usr/local/aws-cli -b /usr/local/bin && \
 	rm -r $tmpdir
+
+COPY --from=cosign-bin \
```
Comment (developer-guy) [inline images/docker-dind/Dockerfile:52]: great

File: images/build-plugins/Makefile (hunk 4 above)
Comment (maxgio92) [reply images/build-plugins/Makefile:4]: Thank you @developer-guy!

File: images/build-libs/Makefile (hunk 3 above)
Comment (maxgio92) [reply images/build-libs/Makefile:26]: Thank you @developer-guy! I think this image (`build-libs`) image and related stuff can be removed, now that the CI has been moved to Github Actions.

File: config/prow/images-publish-rbac.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
```
Comment (maxgio92) [inline config/prow/images-publish-rbac.yaml:1]: The ProwJobs that publish the infra images, will need to access the `Secret` that contain the key needed for the signin.

File: tools/deploy_prow.sh [hunk 9]
```diff
@@ -47,6 +47,12 @@ function launchProwConfig() {
   # PR Status
   kubectl create secret generic github-oauth-config --from-literal=secret="$(./tools/1password.sh -d prow-prstatus-github-oauth-app.yaml)" || true
   kubectl create secret generic cookie --from-literal=secret="$(./tools/1password.sh -d prow-prstatus-cookie-encryption-key.txt)" || true
+
```
Comment (maxgio92) [inline tools/deploy_prow.sh:50]: Despite `cosign` is used with this PR with Kubernetes Secrets as backend, we keep 1Password as the source of truth for secrets (still, we need to better automate or change system).

Comment (maxgio92) [reply tools/deploy_prow.sh:50]: cc @zuc

**Comment by developer-guy** [comment]
Is there anything that I can help with, please limme know if so?

**Comment by maxgio92** [comment]
Thanks @developer-guy. I need to rebase the PR and then I think it could be merged.

The used backend here is the simplest, with Kubernetes and it's the starting point. It's good enough I think, also because we need a global working group to establish a signing architecture for the whole Falco artifacts. WDYT @developer-guy ?

**Comment by developer-guy** [comment]
count me in 🙋🏻‍♂️

**Comment by leogr** [comment]
@maxgio92 is this still a thing? :thinking:

**Comment by maxgio92** [comment]
Thanks @leogr for the catch. I think this can be closed.

Also, because cosign can be used keyless.

**Comment by leogr** [comment]
> Thanks @leogr for the catch. I think this can be closed.
> 
> Also, because cosign can be used keyless.

Agree. Closing. Feel free to reopen it, if any compelling reason comes up :+1: 
/close

**Pull request closed without merging**